rides-and-fares/src/main/java/org/apache/flink/training/exercises/ridesandfares/RidesAndFaresExercise.java

```diff
@@ -18,68 +18,119 @@
+        rides.connect(fares)
+                .flatMap(new EnrichmentFunction())
+                .uid("enrichment") // uid for this operator's state
+                .name("enrichment") // name for this operator in the web UI
+                .addSink(sink);
+
+        // Execute the pipeline and return the result.
+        return env.execute("Join Rides with Fares");
+    }
+
+    /** Creates and executes the pipeline using the default StreamExecutionEnvironment. */
+    public JobExecutionResult execute() throws Exception {
         StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
-        env.setParallelism(ExerciseBase.parallelism);
 
-        DataStream<TaxiRide> rides =
-                env.addSource(rideSourceOrTest(new TaxiRideGenerator()))
-                        .filter((TaxiRide ride) -> ride.isStart)
-                        .keyBy((TaxiRide ride) -> ride.rideId);
+        return execute(env);
+    }
+
+    /**
+     * Main method.
+     *
+     * @throws Exception which occurs during job execution.
+     */
+    public static void main(String[] args) throws Exception {
 
-        DataStream<TaxiFare> fares =
-                env.addSource(fareSourceOrTest(new TaxiFareGenerator()))
-                        .keyBy((TaxiFare fare) -> fare.rideId);
+        RidesAndFaresExercise job =
+                new RidesAndFaresExercise(
+                        new TaxiRideGenerator(),
+                        new TaxiFareGenerator(),
+                        new PrintSinkFunction<>());
 
-        DataStream<Tuple2<TaxiRide, TaxiFare>> enrichedRides =
-                rides.connect(fares).flatMap(new EnrichmentFunction());
+        // Setting up checkpointing so that the state can be explored with the State Processor API.
+        // Generally it's better to separate configuration settings from the code,
+        // but for this example it's convenient to have it here for running in the IDE.
+        Configuration conf = new Configuration();
+        conf.setString("state.backend", "filesystem");
+        conf.setString("state.checkpoints.dir", "file:///tmp/checkpoints");
+        conf.setString("execution.checkpointing.interval", "10s");
+        conf.setString(
+                "execution.checkpointing.externalized-checkpoint-retention",
+                "RETAIN_ON_CANCELLATION");
 
-        printOrTest(enrichedRides);
+        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment(conf);
 
-        env.execute("Join Rides with Fares (java RichCoFlatMap)");
+        job.execute(env);
     }
 
     public static class EnrichmentFunction
-            extends RichCoFlatMapFunction<TaxiRide, TaxiFare, Tuple2<TaxiRide, TaxiFare>> {
+            extends RichCoFlatMapFunction<TaxiRide, TaxiFare, RideAndFare> {
 
         @Override
-        public void open(Configuration config) throws Exception {
+        public void open(Configuration config) throws MissingSolutionException {
```

Comment:
Why not use `Exception` here? For the exercise skeleton, we should keep the original signatures (even though they may generate IntelliJ warnings like in `flatMap1`,... because we are not using them.
I'd like to not scare away beginners from throwing exceptions (should anything be wrong) or using library code that throws exceptions.